### PR TITLE
feat(explorer): search results experiment type filtering

### PR DESCRIPTION
**Pull request by noctillion**
This PR add filter by experiment types in experiment table in explorer section. Allow the export of filtered data to csv

### CODE REVIEW

File: src/components/explorer/searchResultsTables/ExperimentsTable.js [hunk 1]
```diff
@@ -73,14 +73,27 @@ const ExperimentsTable = ({ data, datasetID }) => {
         (state) => state.explorer.tableSortOrderByDatasetID[datasetID]?.["experiments"],
     );
 
+    const experimentTypeFilters = useDynamicFilterOptions(data, "experimentType");
+
+    const filterColumns = SEARCH_RESULT_COLUMNS_EXP.map((col) => {
+        if (col.dataIndex === "experimentType") {
```
Comment (davidlougheed) [inline src/components/explorer/searchResultsTables/ExperimentsTable.js:79]: this is a confusing way to alter `SEARCH_RESULT_COLUMNS_EXP`. It'd be better to move the columns definition into a new memoized variable which has dependencies on `experimentTypeFilters`.

File: src/components/explorer/hooks/explorerHooks.js [hunk 2]
```diff
@@ -29,3 +29,13 @@ export const useSortedColumns = (data, tableSortOrder, columnsDefinition) => {
 
     return { sortedData, columnsWithSortOrder };
 };
+
+export const useDynamicFilterOptions = (data, key) => {
```
Comment (davidlougheed) [inline src/components/explorer/hooks/explorerHooks.js:33]: mention in hook name that this is for tables, like `useDynamicTableFilterOptions`

File: src/components/explorer/ExplorerSearchResultsTable.js [hunk 3]
```diff
@@ -159,6 +175,7 @@ const ExplorerSearchResultsTable = ({
                     pagination={{
                         pageSize: PAGE_SIZE,
                         defaultCurrent: currentPage,
+                        showSizeChanger: false,
```
Comment (davidlougheed) [inline src/components/explorer/ExplorerSearchResultsTable.js:178]: why is this changed?

File: src/components/explorer/ExplorerSearchResultsTable.js [hunk 4]
```diff
@@ -83,8 +99,8 @@ const ExplorerSearchResultsTable = ({
                 key: "all-data",
                 text: "Select All Data",
                 onSelect: () => {
-                    const allRowKeys = data.map((item) => item.key);
-                    handleSetSelectedRows(allRowKeys);
+                    const filteredKeys = filteredData.map((item) => item.key);
```
Comment (davidlougheed) [inline src/components/explorer/ExplorerSearchResultsTable.js:102]: memo dependency should be on filteredData now, not data

File: src/components/explorer/ExplorerSearchResultsTable.js [hunk 5]
```diff
@@ -63,8 +77,10 @@ const ExplorerSearchResultsTable = ({
     }, [dispatch, dataset, activeTab]);
 
     const onPageChange = useCallback((pageObj, filters, sorter) => {
```
Comment (davidlougheed) [inline src/components/explorer/ExplorerSearchResultsTable.js:79]: is this called on filter change too?  in which case maybe we should call this `onTableChange` instead

File: src/components/explorer/ExplorerSearchResultsTable.js [hunk 6]
```diff
@@ -29,11 +29,25 @@ const ExplorerSearchResultsTable = ({
 }) => {
     const { dataset } = useParams();
     const [currentPage, setCurrentPage] = useState(initialCurrentPage || 1);
-    const [numResults] = useState(data.length);
+    const [activeFilter, setActiveFilter] = useState([]);
```
Comment (davidlougheed) [inline src/components/explorer/ExplorerSearchResultsTable.js:32]: can this be multiple things? should it be `activeFilters` instead?

**Review by davidlougheed — CHANGES_REQUESTED** [review]

File: src/components/explorer/ExplorerSearchResultsTable.js (hunk 5 above)
Comment (noctillion) [reply src/components/explorer/ExplorerSearchResultsTable.js:79]: yes, the onPageChange function is called when filters change, in addition to handling pagination changes

File: src/components/explorer/ExplorerSearchResultsTable.js (hunk 4 above)
Comment (noctillion) [reply src/components/explorer/ExplorerSearchResultsTable.js:102]: good catch! Thanks

File: src/components/explorer/ExplorerSearchResultsTable.js [hunk 7]
```diff
@@ -29,11 +29,25 @@ const ExplorerSearchResultsTable = ({
...
-    const [numResults] = useState(data.length);
+    const [activeFilters, setActiveFilters] = useState([]);
 
     const [summaryModalVisible, setSummaryModalVisible] = useState(false);
     const [tracksModalVisible, setTracksModalVisible] = useState(false);
 
+    const filteredData = useMemo(() => {
+        return data.filter((item) => {
+            return Object.keys(activeFilters).every((filterKey) => {
+                const filterValues = activeFilters[filterKey] || [];
```
Comment (davidlougheed) [inline src/components/explorer/ExplorerSearchResultsTable.js:40]: object.keys + every + a prop access is a confusing way to write this, when it could just be `Object.entries(activeFilters).every(([filterKey, filterValues]) => ...)`. Also, I'd prefer expression-based functions over the multiple-nested-closure with return style seen here.

File: src/components/explorer/ExplorerSearchResultsTable.js [hunk 8]
```diff
@@ -29,11 +29,25 @@ const ExplorerSearchResultsTable = ({
...
+                if (filterValues.length === 0) {
+                    return true;
+                }
+                return filterValues.includes(item[filterKey]);
+            });
+        });
+    }, [data, activeFilters]);
+
+    const numResults = useMemo(() => filteredData.length, [filteredData]);
```
Comment (davidlougheed) [inline src/components/explorer/ExplorerSearchResultsTable.js:49]: this memo doesn't do much, since `.length` is a prop of filteredData that doesn't take much time to compute and isn't an object. I'd remove this.

File: src/components/explorer/searchResultsTables/ExperimentsTable.js [hunk 9]
```diff
@@ -28,59 +28,67 @@ ExperimentRender.propTypes = {
...
-
 const ExperimentsTable = ({ data, datasetID }) => {
     const tableSortOrder = useSelector(
         (state) => state.explorer.tableSortOrderByDatasetID[datasetID]?.["experiments"],
     );
 
+    const experimentTypeFilters = useDynamicTableFilterOptions(data, "experimentType");
+
+    const columns = useMemo(() => {
+        const baseColumns = [
```
Comment (davidlougheed) [inline src/components/explorer/searchResultsTables/ExperimentsTable.js:39]: just return this array directly instead of assigning it to a variable which immediately gets returned.

**Review by davidlougheed — CHANGES_REQUESTED** [review]
when I filter and then export CSV without checking any checkbox, I expect to get only the rows I filtered back. instead, I get the whole of the search results back.

**Review by davidlougheed — CHANGES_REQUESTED** [review]

**Comment by davidlougheed** [comment]
otherwise mostly working well!

**Review by davidlougheed — APPROVED** [review]
lgtm
